File: ocaml/planet-time/lib/time_calc.ml
(* time_calc.ml — Planet time calculations for the Interplanetary Time Library
   Story 18.18 — OCaml port of planet-time
   Exact port of planet-time.js / libinterplanet.c time functions. *)

(* ── Planet data helpers ─────────────────────────────────────────────────── *)

(* Map body index to planet data array index (Moon=8 -> Earth=2) *)
let pdata_idx (body : int) : int =
  if body = 8 then 2 else body

(* ── Solar day length ────────────────────────────────────────────────────── *)

(* Solar day length in seconds for a body *)
let solar_day_seconds ~(body : int) : float =
  let idx = pdata_idx body in
  Constants.planets.(idx).Constants.solar_day_ms /. 1000.0

(* ── Local solar time ────────────────────────────────────────────────────── *)

(* Local solar time in seconds since midnight for a body,
   given Julian Day and surface longitude in degrees. *)
let local_solar_time ~(body : int) ~(jd : float) ~(longitude : float) : float =
  let utc_ms = (jd -. Constants.unix_epoch_jd) *. 86400000.0 in
  let idx    = pdata_idx body in
  let pl     = Constants.planets.(idx) in
  (* timezone offset from longitude: longitude/360 * solar_day_ms *)
  let tz_ms = longitude /. 360.0 *. pl.Constants.solar_day_ms in
  let elapsed_ms = utc_ms -. pl.Constants.epoch_ms +. tz_ms in
  let total_days = elapsed_ms /. pl.Constants.solar_day_ms in
  let day_frac   = total_days -. floor total_days in
  day_frac *. pl.Constants.solar_day_ms /. 1000.0

(* ── Sol number ──────────────────────────────────────────────────────────── *)

(* Fractional sol/day number since planet epoch *)
let sol_number ~(body : int) ~(jd : float) : float =
  let utc_ms = (jd -. Constants.unix_epoch_jd) *. 86400000.0 in
  let idx    = pdata_idx body in
  let pl     = Constants.planets.(idx) in
  let elapsed_ms = utc_ms -. pl.Constants.epoch_ms in
  elapsed_ms /. pl.Constants.solar_day_ms

(* ── Planet time record ──────────────────────────────────────────────────── *)

type planet_time = {
  body                      : int;
  jd                        : float;
  sol                       : float;
  local_time_sec            : float;
  day_length_sec            : float;
  light_travel_from_earth_sec : float;
}

(* Compute planet_time record for a body given unix_ms (float) *)
let planet_time ~(body : int) ~(unix_ms : float) : planet_time =
  let jd_val = Orbital.jde unix_ms in
  let sol_val = sol_number ~body ~jd:jd_val in
  let lst = local_solar_time ~body ~jd:jd_val ~longitude:0.0 in
  let day_s = solar_day_seconds ~body in
  (* Light travel from Earth (body=2); Earth itself returns 0 *)
  let lt =
    if body = 2 || body = 8 then 0.0
    else Orbital.light_travel_time_ms ~body1:2 ~body2:body ~utc_ms:unix_ms
  in
  { body; jd = jd_val; sol = sol_val; local_time_sec = lst;
    day_length_sec = day_s; light_travel_from_earth_sec = lt }

(* ── Full planet time (for fixture validation) ───────────────────────────── *)

type full_planet_time = {
  hour            : int;
  minute          : int;
  second          : int;
  local_hour      : float;
  day_fraction    : float;
  day_number      : int;
  day_in_year     : int;
  year_number     : int;
  period_in_week  : int;
  is_work_period  : bool;
  is_work_hour    : bool;
  time_str        : string;
  time_str_full   : string;
  sol_in_year     : int option;
  sols_per_year   : int option;
  zone_id         : string option;
}

(* Zone prefix table: body index → prefix string (Earth=2, Moon=8 are absent) *)
let zone_prefix_of_body (body : int) : string option =
  match body with
  | 0 -> Some "MMT"  (* Mercury *)
  | 1 -> Some "VMT"  (* Venus   *)
  | 3 -> Some "AMT"  (* Mars    *)
  | 4 -> Some "JMT"  (* Jupiter *)
  | 5 -> Some "SMT"  (* Saturn  *)
  | 6 -> Some "UMT"  (* Uranus  *)
  | 7 -> Some "NMT"  (* Neptune *)
  | 8 -> Some "LMT"  (* Moon    *)
  | _ -> None        (* Earth (2) and unknown *)

let get_planet_time ~(body : int) ~(utc_ms : float) : full_planet_time =
  let idx = pdata_idx body in
  let pl  = Constants.planets.(idx) in
  let elapsed_ms  = utc_ms -. pl.Constants.epoch_ms in
  let total_days  = elapsed_ms /. pl.Constants.solar_day_ms in
  let day_number  = int_of_float (floor total_days) in
  let day_frac    = total_days -. floor total_days in
  let local_hour  = day_frac *. 24.0 in
  let h = int_of_float (floor local_hour) in
  let m = int_of_float (floor ((local_hour -. float_of_int h) *. 60.0)) in
  let s = int_of_float (floor (((local_hour -. float_of_int h) *. 60.0
                                 -. float_of_int m) *. 60.0)) in
  (* Mercury/Venus use Earth-clock scheduling: UTC day-of-week + UTC hour *)
  let (piw, is_work_period, is_work_hour) =
    if pl.Constants.earth_clock_sched then begin
      (* dow = ((floor(utc_ms / 86400000) % 7) + 3) % 7, Mon=0..Sun=6 *)
      let utc_day_int = int_of_float (floor (utc_ms /. 86400000.0)) in
      let dow = ((utc_day_int mod 7) + 3 + 7) mod 7 in
      let is_wp = dow < pl.Constants.work_periods_per_week in
      let ms_of_day = int_of_float utc_ms - utc_day_int * 86400000 in
      let utc_h = ms_of_day / 3600000 in
      let is_wh = is_wp
                  && utc_h >= pl.Constants.work_hours_start
                  && utc_h < pl.Constants.work_hours_end in
      (dow, is_wp, is_wh)
    end else begin
      let total_periods = total_days /. pl.Constants.days_per_period in
      let period_int    = int_of_float (floor total_periods) in
      let ppw           = pl.Constants.periods_per_week in
      let piw           = ((period_int mod ppw) + ppw) mod ppw in
      let is_wp = piw < pl.Constants.work_periods_per_week in
      let is_wh = is_wp
                  && local_hour >= float_of_int pl.Constants.work_hours_start
                  && local_hour < float_of_int pl.Constants.work_hours_end in
      (piw, is_wp, is_wh)
    end
  in
  let year_len_days = pl.Constants.sidereal_yr_ms /. pl.Constants.solar_day_ms in
  let year_number   = int_of_float (floor (total_days /. year_len_days)) in
  let day_in_year_f = total_days -. float_of_int year_number *. year_len_days in
  let day_in_year   = int_of_float (floor day_in_year_f) in
  let sol_in_year, sols_per_year =
    if idx = 3 (* Mars *) then begin
      let spy_f = Constants.planets.(3).Constants.sidereal_yr_ms
                  /. Constants.planets.(3).Constants.solar_day_ms in
      (Some day_in_year, Some (int_of_float (floor (spy_f +. 0.5))))
    end else (None, None)
  in
  let zone_id =
    match zone_prefix_of_body body with
    | None        -> None
    | Some prefix -> Some (prefix ^ "+0")
  in
  { hour = h; minute = m; second = s;
    local_hour; day_fraction = day_frac;
    day_number; day_in_year; year_number;
    period_in_week = piw; is_work_period; is_work_hour;
    time_str      = Printf.sprintf "%02d:%02d" h m;
    time_str_full = Printf.sprintf "%02d:%02d:%02d" h m s;
    sol_in_year; sols_per_year; zone_id }

File: lua/planet-time/src/time_calc.lua
-- time_calc.lua — Planet time calculations for interplanet_time Lua library
-- Story 18.19 — exact port of planet-time.js / Python _time.py

local M = {}

local C      = require("src.constants")
local Orbital = require("src.orbital")
local math   = math

-- ── Zone prefix table ─────────────────────────────────────────────────────────
-- Maps planet index to interplanetary timezone prefix string.
-- Earth (2) is absent — returns nil.
local ZONE_PREFIXES = {
  [C.MERCURY] = "MMT",
  [C.VENUS]   = "VMT",
  [C.MARS]    = "AMT",
  [C.JUPITER] = "JMT",
  [C.SATURN]  = "SMT",
  [C.URANUS]  = "UMT",
  [C.NEPTUNE] = "NMT",
  [C.MOON]    = "LMT",
}

-- ── Internal helpers ──────────────────────────────────────────────────────────

-- Return the planet data entry (Moon maps to Earth).
local function planet_data(planet_idx)
  if planet_idx == C.MOON then
    return C.PLANETS[C.EARTH]
  end
  return C.PLANETS[planet_idx]
end

-- ── Public API ────────────────────────────────────────────────────────────────

--- Solar day length in seconds for a given body.
-- @param body_idx  Planet index (0=Mercury..8=Moon)
-- @return seconds
function M.solar_day_seconds(body_idx)
  local p = planet_data(body_idx)
  return p.solar_day_ms / 1000.0
end

--- Local solar time on a planet (seconds since midnight at given longitude).
-- @param body_idx      Planet index
-- @param utc_ms        UTC milliseconds
-- @param longitude_deg Longitude in degrees (east positive; 0 = prime meridian)
-- @return seconds since midnight (0..solar_day_sec)
function M.local_solar_time(body_idx, utc_ms, longitude_deg)
  longitude_deg = longitude_deg or 0.0
  local p = planet_data(body_idx)
  -- Convert longitude fraction of solar day to ms offset
  local tz_adjust_ms = (longitude_deg / 360.0) * p.solar_day_ms
  local elapsed_ms   = (utc_ms - p.epoch_ms) + tz_adjust_ms
  local solar_day_ms = p.solar_day_ms
  -- Day fraction
  local total_days  = elapsed_ms / solar_day_ms
  local day_frac    = total_days - math.floor(total_days)
  return day_frac * (solar_day_ms / 1000.0)
end

--- Sol number (total solar days since planet epoch) at a given UTC time.
-- @param body_idx  Planet index
-- @param utc_ms   UTC milliseconds
-- @return number (fractional)
function M.sol_number(body_idx, utc_ms)
  local p = planet_data(body_idx)
  return (utc_ms - p.epoch_ms) / p.solar_day_ms
end

--- Full planet-time record for a body at a given UTC instant.
-- @param body_idx  Planet index (0=Mercury..8=Moon)
-- @param unix_ms   UTC milliseconds since Unix epoch
-- @return table {
--   body, jd, sol, local_time_sec, day_length_sec,
--   light_travel_from_earth_sec,
--   hour, minute, second, local_hour, day_fraction,
--   day_number, day_in_year, year_number,
--   period_in_week, is_work_period, is_work_hour,
--   time_str, time_str_full,
--   sol_in_year (Mars only), sols_per_year (Mars only),
--   mtc (Mars only)
-- }
function M.planet_time(body_idx, unix_ms)
  -- Moon uses Earth's schedule (tidally locked)
  local key = (body_idx == C.MOON) and C.EARTH or body_idx
  local p   = C.PLANETS[key]

  local elapsed_ms  = unix_ms - p.epoch_ms
  local total_days  = elapsed_ms / p.solar_day_ms
  local day_number  = math.floor(total_days)
  local day_frac    = total_days - day_number

  local local_hour  = day_frac * 24.0
  local h  = math.floor(local_hour)
  local m  = math.floor((local_hour - h) * 60.0)
  local s  = math.floor(((local_hour - h) * 60.0 - m) * 60.0)

  -- Work period / week
  local piw, is_work_period, is_work_hour
  if p.earth_clock_sched then
    -- Mercury and Venus: schedule on UTC Mon–Fri 09:00–17:00
    -- UTC day-of-week: (floor(unix_ms/86400000) % 7 + 3) % 7  →  Mon=0..Sun=6
    piw = ((math.floor(unix_ms / 86400000) % 7) + 3) % 7
    is_work_period = piw < p.work_periods_per_week
    local utc_ms_in_day = unix_ms % 86400000
    local utc_hour = utc_ms_in_day / 3600000.0
    is_work_hour = is_work_period
               and utc_hour >= p.work_hours_start
               and utc_hour < p.work_hours_end
  else
    local total_periods = total_days / p.days_per_period
    local period_int    = math.floor(total_periods)
    piw = ((period_int % p.periods_per_week) + p.periods_per_week) % p.periods_per_week
    is_work_period = piw < p.work_periods_per_week
    is_work_hour   = is_work_period
                  and local_hour >= p.work_hours_start
                  and local_hour < p.work_hours_end
  end

  -- Year / day-in-year
  local year_len_days = p.sidereal_yr_ms / p.solar_day_ms
  local year_number   = math.floor(total_days / year_len_days)
  local day_in_year   = total_days - year_number * year_len_days

  -- Julian Day (TT-based JDE from unix_ms)
  local tai_offset = 10
  for _, entry in ipairs(C.LEAP_SECS) do
    if unix_ms >= entry[2] then tai_offset = entry[1] else break end
  end
  local tt_ms = unix_ms + (tai_offset + 32.184) * 1000.0
  local jd    = C.UNIX_EPOCH_JD + tt_ms / 86400000.0

  -- Light travel time from Earth (skip for Earth/Moon)
  local light_travel = nil
  if body_idx ~= C.EARTH and body_idx ~= C.MOON then
    light_travel = Orbital.light_travel_time(C.EARTH, body_idx, unix_ms)
  end

  -- Mars-specific MTC
  local sol_in_year, sols_per_year, mtc = nil, nil, nil
  if body_idx == C.MARS then
    local sols_per_year_f = C.PLANETS[C.MARS].sidereal_yr_ms / C.PLANETS[C.MARS].solar_day_ms
    sol_in_year  = math.floor(day_in_year)
    sols_per_year = math.floor(sols_per_year_f + 0.5) -- round

    -- MTC (Mars Coordinated Time — sol + time of day from Mars epoch)
    local total_sols_mars = (unix_ms - C.MARS_EPOCH_MS) / C.MARS_SOL_MS
    local mtc_sol  = math.floor(total_sols_mars)
    local mtc_frac = total_sols_mars - mtc_sol
    local mh = math.floor(mtc_frac * 24.0)
    local mm = math.floor((mtc_frac * 24.0 - mh) * 60.0)
    local ms2 = math.floor(((mtc_frac * 24.0 - mh) * 60.0 - mm) * 60.0)
    mtc = { sol = mtc_sol, hour = mh, minute = mm, second = ms2 }
  end

  -- Sol from planet epoch (same as day_number for Mars; total_days for others)
  local sol_total = total_days

  -- Zone ID: PREFIX+0 for non-Earth bodies (no tz_offset_h param in this API)
  local zone_id = nil
  local prefix = ZONE_PREFIXES[body_idx]
  if prefix then
    zone_id = prefix .. "+0"
  end

  return {
    body                      = body_idx,
    jd                        = jd,
    sol                       = sol_total,
    local_time_sec            = day_frac * (p.solar_day_ms / 1000.0),
    day_length_sec            = p.solar_day_ms / 1000.0,
    light_travel_from_earth_sec = light_travel,
    -- time fields
    hour                      = h,
    minute                    = m,
    second                    = s,
    local_hour                = local_hour,
    day_fraction              = day_frac,
    day_number                = day_number,
    day_in_year               = math.floor(day_in_year),
    year_number               = year_number,
    period_in_week            = piw,
    is_work_period            = is_work_period,
    is_work_hour              = is_work_hour,
    time_str                  = string.format("%02d:%02d", h, m),
    time_str_full             = string.format("%02d:%02d:%02d", h, m, s),
    -- Mars only
    sol_in_year               = sol_in_year,
    sols_per_year             = sols_per_year,
    mtc                       = mtc,
    -- Zone ID
    zone_id                   = zone_id,
  }
end

return M
